File: app/routers/content_calendar.py
"""Content calendar router.

Provides unified read/write access to scheduled content items
across posts, broadcasts, and VOD releases.

Registered in app/main.py with prefix="/ui/content-calendar".
"""
from __future__ import annotations

from typing import Literal, Optional

from fastapi import APIRouter, Depends, Query

from app.services.sessions import require_ui_session
from app.services.content_calendar import (
    get_content_calendar,
    get_today_agenda,
    reschedule_item,
    cancel_item,
    ContentType,
)

router = APIRouter(prefix="/ui/content-calendar", tags=["content-calendar"])


@router.get("")
def content_calendar(
    from_ts: int = Query(..., description="Start of time range (Unix seconds)"),
    to_ts: int = Query(..., description="End of time range (Unix seconds)"),
    types: Optional[str] = Query(
        default=None,
        description="Comma-separated content types: post,broadcast,vod",
    ),
    session=Depends(require_ui_session),
):
    """Get all scheduled content for the authenticated creator in a time range."""
    user_id = session["user_sub"]

    type_filter = None
    if types:
        raw = {t.strip().lower() for t in types.split(",")}
        valid: set[str] = set()
        for t in raw:
            if t in ("post", "broadcast", "vod"):
                valid.add(t)
        if valid:
            type_filter = valid

    return get_content_calendar(user_id, from_ts, to_ts, type_filter=type_filter)


@router.get("/today")
def content_calendar_today(
    session=Depends(require_ui_session),
):
    """Get today's and tomorrow's scheduled items."""
    return get_today_agenda(session["user_sub"])


@router.get("/conflicts")
def content_calendar_conflicts(
    from_ts: int = Query(..., description="Start of time range (Unix seconds)"),
    to_ts: int = Query(..., description="End of time range (Unix seconds)"),
    session=Depends(require_ui_session),
):
    """Detect scheduling conflicts in a time range."""
    result = get_content_calendar(session["user_sub"], from_ts, to_ts)
    return {"conflicts": result["conflicts"], "count": len(result["conflicts"])}


@router.post("/reschedule")
def content_calendar_reschedule(
    item_id: str = Query(...),
    item_type: Literal["post", "broadcast", "vod"] = Query(...),
    new_scheduled_at: int = Query(..., description="New scheduled time (Unix seconds)"),
    session=Depends(require_ui_session),
):
    """Reschedule a content item to a new time."""
    return reschedule_item(
        user_id=session["user_sub"],
        item_id=item_id,
        item_type=item_type,
        new_scheduled_at=new_scheduled_at,
    )


@router.post("/cancel")
def content_calendar_cancel(
    item_id: str = Query(...),
    item_type: Literal["post", "broadcast", "vod"] = Query(...),
    session=Depends(require_ui_session),
):
    """Cancel a scheduled content item."""
    return cancel_item(
        user_id=session["user_sub"],
        item_id=item_id,
        item_type=item_type,
    )

File: frontend/src/api/endpoints/content-calendar.ts
import { api } from "@/api/client";
import type {
  ContentCalendarResponse,
  TodayAgendaResponse,
  ConflictsResponse,
  ContentCalendarItem,
  ContentItemType,
} from "@/api/types";

export async function getContentCalendar(
  fromTs: number,
  toTs: number,
  types?: ContentItemType[],
): Promise<ContentCalendarResponse> {
  const params: Record<string, string> = {
    from_ts: String(fromTs),
    to_ts: String(toTs),
  };
  if (types && types.length > 0) {
    params.types = types.join(",");
  }
  return api.get<ContentCalendarResponse>("/ui/content-calendar", params);
}

export async function getTodayAgenda(): Promise<TodayAgendaResponse> {
  return api.get<TodayAgendaResponse>("/ui/content-calendar/today");
}

export async function getConflicts(
  fromTs: number,
  toTs: number,
): Promise<ConflictsResponse> {
  return api.get<ConflictsResponse>("/ui/content-calendar/conflicts", {
    from_ts: String(fromTs),
    to_ts: String(toTs),
  });
}

export async function rescheduleCalendarItem(
  itemId: string,
  itemType: ContentItemType,
  newScheduledAt: number,
): Promise<ContentCalendarItem> {
  return api.post<ContentCalendarItem>(
    "/ui/content-calendar/reschedule",
    null,
    {
      item_id: itemId,
      item_type: itemType,
      new_scheduled_at: String(newScheduledAt),
    },
  );
}

export async function cancelCalendarItem(
  itemId: string,
  itemType: ContentItemType,
): Promise<{ ok: string; id: string; type: ContentItemType }> {
  return api.post<{ ok: string; id: string; type: ContentItemType }>(
    "/ui/content-calendar/cancel",
    null,
    {
      item_id: itemId,
      item_type: itemType,
    },
  );
}
